src/tunnel.cc

```diff
@@ -1311,33 +1303,68 @@ TunnelStateData::saveError(ErrorState *error)
     savedError = error;
 }
 
-/// Starts sending the given error message to the client, leading to the
-/// eventual transaction termination. Call with savedError to send savedError.
+/// If the client expects error message, starts sending the given
+/// error message, leading to the eventual transaction termination.
+/// Otherwise initiates the transaction termination, which will happen
+/// only after all in-progress writing is complete.
 void
-TunnelStateData::sendError(ErrorState *finalError, const char *reason)
+TunnelStateData::sendErrorAndDestroy(ErrorState *err, const char *reason)
 {
     debugs(26, 3, "aborting transaction for " << reason);
 
+    // get rid of any cached error unless that is what the caller is sending
+    if (savedError != err)
+        delete savedError; // may be nil
+    savedError = nullptr;
+
+    if (finalError) {
+        debugs(26, 4, "the final error already exists: " << finalError << ", discarding new error: " << err);
+        if (finalError != err) {
+            // get rid of an error created after finalError
+            delete err; // may be nil
+        }
+        return;
+    }
+
+    finalError = err ? err : new ErrorState(ERR_CANNOT_FORWARD, Http::scInternalServerError, request.getRaw(), al);
+
     if (request)
         request->hier.stopPeerClock(false);
 
     if (opening())
         cancelOpening(reason);
 
-    assert(finalError);
-
-    // get rid of any cached error unless that is what the caller is sending
-    if (savedError != finalError)
-        delete savedError; // may be nil
-    savedError = nullptr;
-
     // we cannot try other destinations after responding with an error
     PeerSelectionInitiator::subscribed = false; // may already be false
 
     *status_ptr = finalError->httpStatus;
-    finalError->callback = tunnelErrorComplete;
-    finalError->callback_data = this;
-    errorSend(client.conn, finalError);
+
+    const auto canSendError = !client.dirty && Comm::IsConnOpen(client.conn) &&
+                              clientExpectsConnectResponse();
+    if (canSendError) {
```

Comment:
If possible, please move the first two conditions inside TunnelStateData::clientExpectsConnectResponse(), eliminating the canSendError variable in this context. In fact, we probably should test connection openness in ClientHttpRequest::clientExpectsConnectResponse() using ClientHttpRequest::clientConnection.

Reply:
I moved the second condition, but it looks like we cannot easily move the 'client.dirty' there.